Add Doge model

LoserCheems opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #35889
Support the Doge-SLM family of small language models.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

to: @ArthurZucker 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

algo-home committed:
Add Doge Model

LoserCheems committed:
Fix code quality

algo-home committed:
Rollback an error commit

algo-home committed:
Fix config for open-source weights

LoserCheems committed:
Revert "Fix config for open-source weights"

This reverts commit 229cdcac10a6a4274d1dd13b729bc14c98eb0c76.

LoserCheems committed:
Merge branch 'huggingface:main' into add-doge-model

LoserCheems committed:
Merge branch 'main' into add-doge-model

Rocketknight1 commented:
Hi @LoserCheems, and thanks for the PR! The model looks cool and I like the paper too, but we're trying to add new models using **modular** in future. You can see a guide [here](https://huggingface.co/docs/transformers/main/en/modular_transformers), and an example modular PR [here](https://github.com/huggingface/transformers/pull/35569/files).

If you write `modular_doge.py` with inheritance then the `configuration_` and `modeling_` files will be auto-generated. This makes the PR much shorter and easier to review.

LoserCheems committed:
Add modular_doge

LoserCheems committed:
Merge branch 'main' into add-doge-model

Rocketknight1 commented:
Hi @losercheems yes, don't worry, it's a new feature so everyone is a bit confused about it! :sweat_smile: 

Your `modular_doge.py` file looks good! The next step is to find code that's copied from other models in `transformers`, and replace that with inheritance. This will make `modular_doge.py` much smaller, but the full `modeling_doge.py` will still be generated without inheritance. You can see some examples in the Qwen2.5 PR:

![image](https://github.com/user-attachments/assets/97ee6a9b-c8d6-449a-b64a-fc5080aa678c)

Classes like `DogeMLP` and `DogeForSequenceClassification` look like they use code from other library classes like Llama, so you could just inherit those instead in the modular file. You can run `make fix-copies` to regenerate `modeling_doge.py` and confirm that it still works.

LoserCheems committed:
Update Doge inherits from Llama

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems commented:
Thank you @Rocketknight1. In fact, because the weight name or config name is different, can directly inherited class is not much, a total of ` RMSNorm `,  ` RotaryEmbedding ` , and ` DogeForSequenceClassification ` inherited from Llama.

LoserCheems committed:
Fix import bug

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
[docs] Add usage of doge model

Rocketknight1 commented:
Hi @losercheems, the last code quality error is caused by an unprotected `import torch`. These need to be guarded by `if is_torch_available` because some people have JAX-only or TF-only systems, and unguarded imports can make it impossible for them to use the library!

There are unrelated failing tests under `tests_torch` - you can ignore them for now, but once you can get code quality green then let me know and I'll review the PR.

LoserCheems committed:
Fix Doge import pretrainedconfig from modeling_utils to configuration_utils

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems commented:
Sorry @Rocketknight1, I mistakenly imported `PretrainedConfig` from `modeling_utils`, which is now fixed.

LoserCheems committed:
[docs] remove trust remote code from doge

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems commented:
em🤓, There seems to be something wrong with `RotaryEmbedding` inherited from Llama.

LoserCheems committed:
Fix dynamo bug in doge model

LoserCheems commented:
gentle ping @Rocketknight1

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Update docstrings

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems commented:
gentle ping @Rocketknight1

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems commented:
@Rocketknight1 code quality is already green!

Rocketknight1 commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1220 @@
+        return causal_mask
+
+
+class KwargsForCausalLM(LossKwargs): ...
```
I think this should probably be imported from Llama, or you should replace the references further down with `LossKwargs`. We don't need to create a class that's just a copy of `LossKwargs`!

Rocketknight1 reviewed (commented):
Hey, I made a couple more comments, and sorry for the delay! I think this is quite close to being ready, but there's some more code that could be removed from the `modular` file, and that'll make it easier for the core maintainers to review later. I've flagged a couple of examples - you can ping me after you clean that up (or if it turns out I'm wrong and you can't import those methods from elsewhere), and I'll get a core maintainer to do a final review

LoserCheems committed:
Import apply_rotary_pos_emb and repeat_kv from Llama

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems commented:
Thank you @Rocketknight1 for your suggestion, I have finished the modification according to the suggestion!

LoserCheems committed:
Merge branch 'main' into add-doge-model

Rocketknight1 commented:
Yes, looks good to me now! cc @cyrilvallez @ArthurZucker for core maintainer review. Failing tests are unrelated, should be fixed by rebasing after review / fixing nits

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+
+        return attn_mask
+
+    def eager_attention_forward(
+        self,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attention_mask: Optional[torch.Tensor],
+        scaling: float,
+        dropout: float = 0.0,
+        **kwargs,
+    ) -> torch.Tensor:
+        key_states = repeat_kv(key, self.num_key_value_groups)
+        value_states = repeat_kv(value, self.num_key_value_groups)
+
+        # compute attention scores matrix
+        attn_weights = torch.matmul(query, key_states.transpose(-1, -2)) * scaling
+        if attention_mask is not None:
+            causal_mask = attention_mask[:, :, :, : key_states.shape[-2]]
+            attn_weights = attn_weights + causal_mask
+
+        # upcast attention scores to fp32
+        attn_weights = F.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+        attn_weights = F.dropout(attn_weights, p=dropout, training=self.training)
+
+        # apply attention scores to value states
+        attn_output = torch.matmul(attn_weights, value_states)
+        attn_output = attn_output.transpose(1, 2).contiguous()
+        return attn_output
```
let's move outside of self for now!

LoserCheems replied:
I moved them outside and used them to update `ALL_ATTENTION_FUNCTIONS`

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+        # get similarity with queries and keys
+        queries = self.queries(hidden_states)
+        queries = queries.view(bsz, seq_len, 2, self.num_cdmoe_heads, -1).permute(2, 0, 1, 3, 4)
+        sim = torch.einsum("p b t h n, h k p n -> p b t h k", queries, self.keys)
+
+        # get experts with the highest similarity
+        (scores_x, scores_y), (indices_x, indices_y) = sim.topk(self.num_cdmoe_experts_per_head, dim=-1)
+        all_scores = scores_x.unsqueeze(-1) + scores_y.unsqueeze(-2)
+        all_scores = all_scores.view(*scores_x.shape[:-1], -1)
+        all_indices = (indices_x.unsqueeze(-1) * self.num_keys) + indices_y.unsqueeze(-2)
+        all_indices = all_indices.view(*indices_x.shape[:-1], -1)
+        scores, pk_indices = all_scores.topk(self.num_cdmoe_experts_per_head, dim=-1)
+        indices = all_indices.gather(-1, pk_indices)
+        down_embed = self.down_embed(indices)
+        up_embed = self.up_embed(indices)
+
+        # mix experts states with cross domain states
+        experts_weights = torch.einsum("b t d, b t h k d -> b t h k", hidden_states, down_embed)
+        experts_weights = self.act_fn(experts_weights) * scores.softmax(dim=-1)
+        experts_states = torch.einsum("b t h k, b t h k d -> b t d", experts_weights, up_embed)
```
we need to get rid of the einsums 😉

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+                `(batch_size, 1, query_length, key_value_length)`.
+            sequence_length (`int`):
+                The sequence length being processed.
+            target_length (`int`):
+                The target length: when generating with static cache, the mask should be as long as the static cache,
+                to account for the 0 padding, the part of the cache that is not filled yet.
+            dtype (`torch.dtype`):
+                The dtype to use for the 4D attention mask.
+            device (`torch.device`):
+                The device to plcae the 4D attention mask on.
+            cache_position (`torch.Tensor`):
+                Indices depicting the position of the input sequence tokens in the sequence.
+            batch_size (`torch.Tensor`):
+                Batch size.
+        """
+        if attention_mask is not None and attention_mask.dim() == 4:
+            # In this case we assume that the mask comes already in inverted form and requires no inversion or slicing.
+            causal_mask = attention_mask
+        else:
+            min_dtype = torch.finfo(dtype).min
+            causal_mask = torch.full(
+                (sequence_length, target_length),
+                fill_value=min_dtype,
+                dtype=dtype,
+                device=device,
+            )
+            if sequence_length != 1:
+                causal_mask = torch.triu(causal_mask, diagonal=1)
+            causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+            causal_mask = causal_mask[None, None, :, :].expand(batch_size, 1, -1, -1)
+            if attention_mask is not None:
+                causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+                mask_length = attention_mask.shape[-1]
+                padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :]
+                padding_mask = padding_mask == 0
+                causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
+                    padding_mask, min_dtype
+                )
+
+        return causal_mask
```
these should be inherited / the same as any other model like Llama 😉

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+    def set_decoder(self, decoder):
+        self.model = decoder
+
+    @deprecate_kwarg("num_logits_to_keep", version="4.50", new_name="logits_to_keep")
```
```suggestion
```

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+    @deprecate_kwarg("num_logits_to_keep", version="4.50", new_name="logits_to_keep")
+    @add_start_docstrings_to_model_forward(DOGE_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=CausalLMOutputWithPast, config_class=_CONFIG_FOR_DOC)
+    def forward(
```
function should be 1-1 the same as Llama's forward !

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+    """,
+    DOGE_START_DOCSTRING,
+)
+class DogeForSequenceClassification(LlamaForSequenceClassification):
```
nice here!

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+        )
+
+
+class RMSNorm(LlamaRMSNorm):
+    def __init__(self, hidden_size, eps=1e-6):
+        """
+        RMSNorm is equivalent to T5LayerNorm
+        """
+        super().__init__(hidden_size, eps)
```
should not be needed, a good default in the config is enough if that is the only diff
```suggestion
class DogeRMSNorm(LlamaRMSNorm):
    pass
```

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+        super().__init__(hidden_size, eps)
+
+
+class Residual(nn.Module):
```
```suggestion
class DogeResidual(nn.Module):
```

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+        return f"{tuple(self.weight.shape)}"
+
+
+class RotaryEmbedding(LlamaRotaryEmbedding):
```
```suggestion
class DogeRotaryEmbedding(LlamaRotaryEmbedding):
```

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+        self.attention_dropout = config.attention_dropout
+        self.dynamic_mask_ratio = config.dynamic_mask_ratio
+
+        self.ALL_ATTENTION_FUNCTIONS = {
+            "eager": self.eager_attention_forward,
+            "flex_attention": self.flex_attention_forward,
+            "sdpa": self.sdpa_attention_forward,
+        }
```
if it is only in self, it's gonna be hard to modify outside the scope of this model only! If someone wants to add "my_custom_function" he is gonna have a hard time!

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+            return score
+
+        mask_mod = causal_mod if self.is_causal else dynamic_mod
+
```
dropout missing from this one, but I guess it's just not really supported

LoserCheems replied:
yes, flex_attention is not support dropout yet...

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+
+        return attn_mask
+
+    def eager_attention_forward(
```
the three attention forward look general (similar to Llama and the ones in `integration/flex` and `integrations/sdpa`. And as my comment above mentioned, you won't have flash if you use `self.ATTENTION` instead of the global scoped one!

ArthurZucker replied:
(Making it incompatible with TGI / vllm 😉 )

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
+
+        # calculate dynamic mask from value_states
+        dt_states = self.dt_proj(
+            value_states.transpose(1, 2).reshape(value_states.shape[0], value_states.shape[-2], -1)
+        )
```
I think this layer can be added to the TP plan as well, if we handle it properly!  SHould be a `rowise`. We need to be a bit careful with the shapes tho!

ArthurZucker commented on src/transformers/models/doge/configuration_doge.py:
```diff
@@ -0,0 +1,228 @@
+        "layers.*.self_attn.q_proj": "colwise",
+        "layers.*.self_attn.k_proj": "colwise",
+        "layers.*.self_attn.v_proj": "colwise",
+        "layers.*.self_attn.dt_proj": "colwise",
```
see my comment afterwards!

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1172 @@
+        # get similarity with queries and keys
+        queries = self.queries(hidden_states)
+        queries = queries.view(bsz, seq_len, 2, self.num_cdmoe_heads, -1).permute(2, 0, 1, 3, 4)
+        sim = torch.einsum("p b t h n, h k p n -> p b t h k", queries, self.keys)
```
what's SIM?

LoserCheems replied:
it is routing weight and I will rename it to `routing_weights`

ArthurZucker reviewed (commented):
Nice work!

algo-home committed:
Fix all nits

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix code quality

algo-home committed:
Fix some bugs

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix code quality

LoserCheems commented:
Thank you @ArthurZucker for your careful review. I have finished the modification according to your suggestion.

LoserCheems committed:
Remove inherited `_update_causal_mask` from Llama
This leads to incorrect weight initialization.

LoserCheems committed:
Fix the wrong tensor orderings in DogeCDMoE

LoserCheems committed:
Merge branch 'main' into add-doge-model

ArthurZucker commented:
Cool having a look!

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Merge branch 'huggingface:main' into add-doge-model

algo-home committed:
Fix attention mask bug
We have to provide attention_mask for dynamic mask computation

LoserCheems committed:
Merge branch 'huggingface:main' into add-doge-model

LoserCheems committed:
Merge branch 'huggingface:main' into add-doge-model

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1224 @@
+        return attn_mask
+
+
+class DogeMLP(nn.Module):
+    def __init__(self, config: DogeConfig):
+        super().__init__()
+        self.hidden_dim = config.hidden_size
+        self.intermediate_dim = config.intermediate_size
+        self.act_fn = ACT2FN[config.hidden_act]
+
+        self.gate_proj = nn.Linear(self.hidden_dim, self.intermediate_dim, bias=config.hidden_bias)
+        self.up_proj = nn.Linear(self.hidden_dim, self.intermediate_dim, bias=config.hidden_bias)
+        self.down_proj = nn.Linear(self.intermediate_dim, self.hidden_dim, bias=config.hidden_bias)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        **kwargs,
+    ) -> torch.Tensor:
+        hidden_states = self.down_proj(self.act_fn(self.gate_proj(hidden_states)) * self.up_proj(hidden_states))
+        return hidden_states
```
This can be inherited from Llama directly! Maybe just overwrites the layers if the config names change

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1224 @@
+        past_key_value: Optional[Cache] = None,
+        output_attentions: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,  # necessary, but kept here for BC
```
Let's make the `position_embeddings` non-optional here 🤗

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1224 @@
+    etc.)
+
+    This model is also a PyTorch [torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) subclass.
+    Use it as a regular PyTorch Module and refer to the PyTorch documentation for all matter related to general usage
+    and behavior.
+
+    Parameters:
+        config ([`DogeConfig`]):
+            Model configuration class with all the parameters of the model. Initializing with a config file does not
+            load the weights associated with the model, only the configuration. Check out the
+            [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+
+@add_start_docstrings(
+    "The bare Doge Model outputting raw hidden-states without any specific head on top.",
+    DOGE_START_DOCSTRING,
+)
+class DogePreTrainedModel(PreTrainedModel):
+    config_class = DogeConfig
+    base_model_prefix = "model"
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["DogeDecoderLayer"]
+    _skip_keys_device_placement = ["past_key_values"]
+    _supports_sdpa = True
+    _supports_flex_attn = True
+    _supports_cache_class = True
+    _supports_quantized_cache = True
+    _supports_static_cache = True
+
+    def _init_weights(self, module):
+        std = self.config.initializer_range
+        if isinstance(module, (nn.Linear)):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
```
This is not needed, you can directly inherit `DogePreTrainedModel` from the Llama one

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1224 @@
+        Args:
+            attention_mask (`torch.Tensor`):
+                A 2D attention mask of shape `(batch_size, key_value_length)` or a 4D attention mask of shape
+                `(batch_size, 1, query_length, key_value_length)`.
+            sequence_length (`int`):
+                The sequence length being processed.
+            target_length (`int`):
+                The target length: when generating with static cache, the mask should be as long as the static cache,
+                to account for the 0 padding, the part of the cache that is not filled yet.
+            dtype (`torch.dtype`):
+                The dtype to use for the 4D attention mask.
+            device (`torch.device`):
+                The device to plcae the 4D attention mask on.
+            cache_position (`torch.Tensor`):
+                Indices depicting the position of the input sequence tokens in the sequence.
+            batch_size (`torch.Tensor`):
+                Batch size.
+        """
+        if attention_mask is not None and attention_mask.dim() == 4:
+            # In this case we assume that the mask comes already in inverted form and requires no inversion or slicing.
+            causal_mask = attention_mask
+        else:
+            min_dtype = torch.finfo(dtype).min
+            causal_mask = torch.full(
+                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
+            )
+            if sequence_length != 1:
+                causal_mask = torch.triu(causal_mask, diagonal=1)
+            causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+            causal_mask = causal_mask[None, None, :, :].expand(batch_size, 1, -1, -1)
+            if attention_mask is not None:
+                causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+                mask_length = attention_mask.shape[-1]
+                padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :]
+                padding_mask = padding_mask == 0
+                causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
+                    padding_mask, min_dtype
+                )
+
+        return causal_mask
```
If you consider renaming just very few weights (i.e. the norm and ambeddings), all this can be removed, and a very elegant inheritance from Llama can be used instead!

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1224 @@
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # decoder output consists of (dec_features, layer_state, dec_hidden, dec_attn)
+        outputs = self.model(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            cache_position=cache_position,
+            **kwargs,
+        )
+
+        hidden_states = outputs[0]
+        # only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
+        logits = self.lm_head(hidden_states[:, slice_indices, :])
+
+        loss = None
+        if labels is not None:
+            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.vocab_size, **kwargs)
+
+        if not return_dict:
+            output = (logits,) + outputs[1:]
+            return (loss,) + output if loss is not None else output
+
+        return CausalLMOutputWithPast(
+            loss=loss,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
```
Here this can all be inherited directly from Llama as well!

LoserCheems replied:
What should I do if I inherit llama's forward method with the wrong example?
 ```python
        >>> from transformers import AutoTokenizer, DogeForCausalLM

        >>> model = DogeForCausalLM.from_pretrained("meta-doge/Doge-2-7b-hf")
        >>> tokenizer = AutoTokenizer.from_pretrained("meta-doge/Doge-2-7b-hf")
```

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,1224 @@
+        )
+
+
+@add_start_docstrings(
+    """
+    The Doge Model transformer with a sequence classification head on top (linear layer).
+
+    [`DogeForSequenceClassification`] uses the last token in order to do the classification, as other causal models
+    (e.g. GPT-2) do.
+
+    Since it does classification on the last token, it requires to know the position of the last token. If a
+    `pad_token_id` is defined in the configuration, it finds the last token that is not a padding token in each row. If
+    no `pad_token_id` is defined, it simply takes the last value in each row of the batch. Since it cannot guess the
+    padding tokens when `inputs_embeds` are passed instead of `input_ids`, it does the same (take the last value in
+    each row of the batch).
+    """,
+    DOGE_START_DOCSTRING,
+)
```
You don't need that, it will be added automatically by modular!

Cyrilvallez reviewed (commented):
Alright, here is a new round of reviews! Basically, a lot of code can be cut by inheriting from Llama classes 🤗 The model itself would need to change some weights keys, but very few of them. We are happy to include a nice weight converting script such as can be found [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/glm/convert_glm_weights_to_hf.py) to the library as well, to facilitate the weight name mapping if needed! 

I also feel like quite a lot of config arguments could be renamed to what already exist in the library: consistency between models ensure that every user of your model will very quickly understand what's happening with your model and config! 
If needed, we are also happy to help with the conversion of your weight name and config args 🤗 But if you don't have that many checkpoints on the hub already, this is usually a pretty straightforward operation!

EDIT: Renaming weights also enables to use modular much more in this case, which in turn means your model will automatically, and 0-day benefit from all refactors/new features in the future!

LoserCheems committed:
Merge branch 'huggingface:main' into add-doge-model

LoserCheems committed:
Merge branch 'huggingface:main' into add-doge-model

LoserCheems committed:
Modify most implementations to inherit from Llama
But there are two problems:
1. `flex_attention_forward` is not updated properly
2. `Example` error in the forward method of DogeForCausalLM

LoserCheems committed:
Modify CDMoE for batch efficient implementation

LoserCheems committed:
Uniform MoE configuration names, just like QwenMoE

LoserCheems committed:
Fix code quality

LoserCheems committed:
Fix code quality

LoserCheems committed:
Fix code quality

LoserCheems committed:
Add tp plan of CDMoE Module

LoserCheems committed:
Hybird DMA with sliding window

LoserCheems committed:
Update valid tokens greater than window size

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix code quality

LoserCheems committed:
Add `convert_doge_weights_to_hf`

LoserCheems commented:
@Cyrilvallez, I have finished the `convert_weights_to_hf` script.

LoserCheems committed:
Fix STATE_DICT_MAPPING in convert_doge_weights_to_hf.py

LoserCheems committed:
Fix nits in modular_doge

LoserCheems commented:
@Cyrilvallez Request a review. There should be no issues other than the example error, but the impact should be minor as I found the same bug for all models using `modular`.

LoserCheems committed:
Fix code quality

Cyrilvallez reviewed (commented):
Alright! Very nice! Still a few comments, but we are extremely close! Nice work 🤗
Make sure to rebase as well to integrate latest changes and fix the conflicts!

LoserCheems committed:
Merge remote-tracking branch 'origin/main' into add-doge-model

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems commented on src/transformers/models/doge/configuration_doge.py:
```diff
@@ -0,0 +1,226 @@
+    model_type = "doge"
+    keys_to_ignore_at_inference = ["past_key_values"]
+    # Default tensor parallel plan for base model `DogeModel`
+    base_model_tp_plan = {
+        "layers.*.self_attn.q_proj": "colwise",
+        "layers.*.self_attn.k_proj": "colwise",
+        "layers.*.self_attn.v_proj": "colwise",
+        "layers.*.self_attn.dt_proj": "rowwise",
+        "layers.*.self_attn.o_proj": "rowwise",
+        "layers.*.mlp.gate_proj": "colwise",
+        "layers.*.mlp.up_proj": "colwise",
+        "layers.*.mlp.down_proj": "rowwise",
+        "layers.*.feed_forward.router_gate": "colwise",
+        "layers.*.mlp.down_embed": "rowwise",
+        "layers.*.mlp.up_embed": "rowwise",
+    }
```
After many tests, `router_gate` does need to be changed from `colwise` to `colwise_rep` to be consistent with mixtral, and the rest of tp_plan is OK

LoserCheems committed:
Fix all nits

LoserCheems commented:
Thank you @Cyrilvallez for your careful review, I have fixed all issues as suggested.

LoserCheems committed:
Fix all nits

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,650 @@
+    pass
+
+
+def flex_attention_forward(
```
do we add this because of softcap? or why are we adding this?

LoserCheems replied:
Because the mask here is **num_heads * q_len * k_len**, So we need to update line 298 from `score = score + causal_mask[batch_idx][0][q_idx][kv_idx]` to `score = score + causal_mask[batch_idx][head_idx][q_idx][kv_idx]`

vasqu replied:
In general: Do we have different score mods for different heads? The initial idea here was that the heads share the same score mods across heads so the head idx didn't matter. Open to have it changed here tho. Also, I'm pretty sure the new mask API will return the block mask only so the score mod should not come into play here :eyes: 

> WDYT about adding support to pass scoremod?

In general, iirc torch people always recommend the block mask for perf. So we should make sure that both would be passed (score mod + block mask) in those cases. Unsure which models would be affected / benefit from it tho - maybe some certain biases like t5(?)

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,650 @@
+        self.num_keys = int(math.sqrt(self.num_experts))
+
+        # cross domain
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=config.hidden_bias)
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=config.hidden_bias)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=config.hidden_bias)
```
in general we separate this one

ArthurZucker replied:
but no worries

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
Very nice! Much aligned with what we have, makes it easy to review thanks for your hard work! 🤗

ArthurZucker commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,650 @@
+    return attn_output, attention_weights
+
+
+ALL_ATTENTION_FUNCTIONS.update({"flex_attention": flex_attention_forward})
```
I don't remember but let's make sure this only updates for this model particularly!

ArthurZucker reviewed (approved):
Looks great! Thanks for aligning 🤗 Ping @Cyrilvallez he'll merge it !

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Make sure the attention function is updated inside the class

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix code quality issues in the Doge model and add a test for it

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix `test_generate`

LoserCheems committed:
Fix code quality

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,651 @@
+from ...integrations.flex_attention import compile_friendly_flex_attention
+from ...modeling_attn_mask_utils import AttentionMaskConverter
+from ...modeling_rope_utils import rope_config_validation
+from ...modeling_utils import ALL_ATTENTION_FUNCTIONS
```
Here, this is the proper way to do it if we need to overwrite an existing attention function!

```suggestion
from ...modeling_utils import AttentionInterface
ALL_ATTENTION_FUNCTIONS = AttentionInterface()
```

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,651 @@
+
+
+class DogeAttention(nn.Module):
+    DOGE_ATTENTION_FUNCTIONS = dict(ALL_ATTENTION_FUNCTIONS)
+    DOGE_ATTENTION_FUNCTIONS.update({"flex_attention": flex_attention_forward})
```
To remove, see previous comments 🤗

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,651 @@
+    # lse is returned in float32
+    attention_weights = attention_weights.to(value.dtype)
+    attn_output = attn_output.transpose(1, 2).contiguous()
+
+    return attn_output, attention_weights
```
And here we need to add it like that!

```suggestion

    return attn_output, attention_weights
    
ALL_ATTENTION_FUNCTIONS["flex_attention"] = flex_attention_forward
```

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,651 @@
+                    'eager attention. This warning can be removed using the argument `attn_implementation="eager"` when loading the model.'
+                )
+            else:
+                attention_interface = self.DOGE_ATTENTION_FUNCTIONS[self.config._attn_implementation]
```
```suggestion
                attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
```

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,651 @@
+        self.top_k = config.num_experts_per_tok
+        self.num_keys = int(math.sqrt(self.num_experts))
+
+        # cross domain
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=config.hidden_bias)
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=config.hidden_bias)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=config.hidden_bias)
```
Here would it make sense to declare it as an MLP instead to simplify? 🤗

Cyrilvallez reviewed (commented):
Hey! Super nice! Just left some final thoughts! 🤗

LoserCheems committed:
Fix nits fllowing suggestions

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix code quality

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems commented:
gentle ping @Cyrilvallez

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix code quality issues

LoserCheems committed:
Fix nits

LoserCheems committed:
Fix code quality nits

LoserCheems committed:
Fix the missing parameters in the configuration.

LoserCheems committed:
Fix the missing parameters in the configuration.

LoserCheems committed:
Fix nits

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,853 @@
+class DogeRMSNorm(LlamaRMSNorm):
+    pass
+
+
+ALL_LAYERNORM_LAYERS.append(DogeRMSNorm)
```
We don't need this one, ket's remove it! 🤗

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,853 @@
+    "DogeConfig",
+    "DogeForCausalLM",
+    "DogeModel",
+    "DogePreTrainedModel",  # noqa: F822
```
No need for the "noqa" comment here!

Cyrilvallez commented on src/transformers/models/doge/modular_doge.py:
```diff
@@ -0,0 +1,853 @@
+
+        return outputs
+
+
+class DogePreTrainedModel(LlamaPreTrainedModel):
```
We need to overwrite the `_init_weights` in this class in order to correctly init the `A` layer in the `Attention` as 0s, as well as the `input_residual` and `post_attention_residual` in the Layers as 1s!

Cyrilvallez reviewed (commented):
Hey @LoserCheems! Super super sorry about the delay! Just some very small changes, then we can merge! 🤗🤗

LoserCheems committed:
Merge branch 'huggingface:main' into add-doge-model

LoserCheems committed:
Add initialization of attention

LoserCheems commented:
Hey @Cyrilvallez, thank you very much for your long review. We've finally completed it!

LoserCheems committed:
Fix last nits

LoserCheems committed:
Simplify dynamic mask generation logic

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Rename router_logits to gate_logits for matching latest changes of MixtralModel

LoserCheems committed:
Rename typings for matching latest changes of MixtralModel

LoserCheems committed:
Fixes typo in comment

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Merge branch 'main' into add-doge-model

ArthurZucker reviewed (approved):
Let's merge!

ArthurZucker committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Update src/transformers/models/doge/modular_doge.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix code quality issues to match other modular

LoserCheems committed:
Merge branch 'add-doge-model' of https://github.com/LoserCheems/transformers into add-doge-model

LoserCheems committed:
Fix code quality issues to match other modular

LoserCheems committed:
Fix the static compilation errors

LoserCheems committed:
Update model weights link

LoserCheems committed:
Merge branch 'main' into add-doge-model

LoserCheems committed:
Fix code quality issues to match other modular

LoserCheems commented:
Hi @ArthurZucker, all the tests have been passed!

LoserCheems committed:
Merge branch 'main' into add-doge-model

Cyrilvallez committed:
Merge branch 'main' into add-doge-model

Cyrilvallez committed:
Merge branch 'main' into add-doge-model

Cyrilvallez committed:
reapply modular and support for new outputs

Cyrilvallez committed:
style

Cyrilvallez committed:
simplify a lot

Cyrilvallez committed:
fix import location

Cyrilvallez committed:
Merge branch 'main' into add-doge-model

Cyrilvallez committed:
reapply modular

Cyrilvallez committed:
fix

Cyrilvallez commented:
run-slow: doge

Cyrilvallez committed:
fix integration test

Cyrilvallez reviewed (approved):
All right! I fixed myself what was missing based on our most recent refactors! This is now good to go, merging! 
Thanks again for your great work, and for bearing with us! 🤗🚀
